File: python/2958.length_of_longest_subarray_with_at_most_k_frequency.py
from collections import defaultdict
from typing import List


class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:

        if len(nums) == 1:
            return 1

        counts = defaultdict(int)
        result = 0
        size = 0
        left = 0

        for _, value in enumerate(nums):
            counts[value] += 1
            size += 1
            while counts[value] > k:
                left_val = nums[left]
                counts[left_val] -= 1
                left += 1
                size -= 1
            result = max(result, size)
        return result

File: cpp/2958.length_of_longest_subarray_with_at_most_k_frequency.cpp
#include <unordered_map>

class Solution {
 public:
    int maxSubarrayLength(const std::vector<int>& nums, int k) {
        int result = 0;
        int left = 0;
        int size = 0;

        std::unordered_map<int, int> counts;

        for (auto num : nums) {
            ++counts[num];
            ++size;

            while (counts[num] > k) {
                int l_num = nums[left];
                --counts[l_num];
                --size;
                ++left;
            }
            result = std::max(result, size);
        }
        return result;
    }
};
